Cross-user Auth (#454)
ietf-wg-gnap/gnap-core-protocol: draft-ietf-gnap-core-protocol.md

    happens out of band from the protocol, such as over the phone between parties. Note that the
    RO is not interacting with the client instance.

2. The end user communicates the identifier to the client instance. The means by which the
    identifier is communicated to the client instance is out of scope for this specification.

3. The client instance [requests access to subject information](#request). The request includes
    the RO's identifier in the [subject information request](#request-subject) `sub_ids` field,
    and the end user's identifier in the [user information field](#request-user) of the request.
    The request includes no interaction start methods, since the end user is not expected to
    be the one interacting with the AS. The request does include the
    [push based interaction finish method](#request-interact-callback-push) to allow the AS
    to signal to the client instance when the interaction with the RO has concluded.

4. The AS sees that the identifier for the end user and subject being requested are different.
    The AS determines that it can reach out to the RO asynchronously for approval. While it

>     is doing so, the AS returns a [continuation response](#response-continue) with a `wait` field
>     to allow the client instance to keep polling.

Author: ```suggestion
    is doing so, the AS returns a [continuation response](#response-continue) with a `finish` nonce
    to allow the client instance to keep polling after interaction with the RO has concluded.
```

5. The AS contacts the RO and has them authenticate to the system. The means for doing this are
    outside the scope of this specification, but the identity of the RO is known from the subject
    identifier sent in (3).

6. The RO is prompted to authorize the end user's request via the client instance. Since the end
    user was identified in (3) via the user field, the AS can show this information to the
    RO during the authorization request.

7. The RO completes the authorization with the AS. The AS marks the request as _approved_.

> 8. The RO pushes the [interaction finish message](#interaction-pushback) to the client instance.

Author: ```suggestion
8. The RO pushes the [interaction finish message](#interaction-pushback) to the client instance.
    Note that in the case the RO cannot be reached or the RO denies the request, the AS still sends the interaction
    finish message to the client instance, after which the client instance can negotiate next steps if possible.
```

    assertion formats are defined by the [Assertion Formats Registry](#IANA-assertion-formats).
    REQUIRED if assertions are requested.

`sub_ids` (array of objects):
: An array of subject identifiers representing the subject that information
    is being requested for. Each object is a subject identifier as defined by

>     {{I-D.ietf-secevent-subject-identifiers}}. All identifiers MUST identify

Author: ```suggestion
    {{I-D.ietf-secevent-subject-identifiers}}. All identifiers in the `sub_ids` array MUST identify
```

release, but such advanced use cases have some distinct issues that implementors need to be fully
aware of before using these features.

First, in many applications, the return of a subject information to the client instance could
indicate to the client instance that the end-user is the party represented by that information,
functionally allowing the end-user to authenticate to the client application. While the details of
a fully functional authentication protocol are outside the scope of GNAP, it is a common
exercise for a client instance to be requesting information about the end user. This is facilitated
by the several [interaction methods](#interaction-start) defined in GNAP that allow the end user
to begin interaction directly with the AS. However, when the subject of the information is
intentionally not the end-user, the client application will need some way to differentiate between
requests for authentication of the end user and requests for information about a different user.
Confusing these states could lead to an attacker having their account associated with a privileged
user. Client instances can mitigate this by having distinct code paths for primary end user

> authentication and requesting subject information about secondary users, such as in a call center.

Reviewer: a typical use case is the support of a client application (e.g. mobile), while the agent uses a corporate PC (e.g. desktop app). So in some cases, the client class_id might help too

Second, RO's interacting asynchronously do not usually have the same context as an end user in an
application attempting to perform the task needing authorization. As such, the asynchronous requests
for authorization coming to the RO from the AS might have very little to do with what the RO is

> doing at the time. This situation can consequently lead to authorization fatigue on the part of the

Reviewer: yes, although authorization fatigue could also happen when done synchronously (people might not read what they approve)

Author: I think that's a valid concern overall, but at least when done synchronously the end-user stands a chance of having enough context to make the right decision.
